[dependency] UI dependency update

veronikaslc opened the pull request:
Updated dependencies:

`React` from v. 18 to v. 19 
`react-router` from v. 5 to v. 7
`node` from v. 16 to v. 22
`yarn`
`webpack`
`babel`
`@mui/x-date-pickers` from v. 6 to v. 7
`@mui/lab` from v. 5 to v. 6
`@mui/icons-material` from v. 5 to v. 6
`@mui/material` from v. 5 to v. 6 
`@mui/styles` from v. 5 to v. 6
`@uiw/react-md-editor` from v. 3 to v. 4
`css-loader` from v. 6 to v. 7
`eslint` from v. 8 to v. 9
`history` from v. 4 to v. 5
`react-to-print` from v. 2 to v. 3
`style-loader` from v. 3 to v. 4
`uuid` from v. 9 to v. 11
`yup` from v. 0 to v. 1
`eslint-plugin-react`
`recharts`
`cornerstone-wado-image-loader`
`react-number-format`
`react-google-autocomplete`
`google-palette`
 
Removed unused dependencies: `semantic-ui-css`, `semantic-ui-react`, `set-blocking`, `formik-material-fields`

veronikaslc committed:
[dependency] Update @mui/x-date-pickers from v. 6 to v. 7

veronikaslc committed:
[dependency] Update @mui/lab from v. 5 to v. 6

veronikaslc committed:
[dependency] Update @mui/icons-material from v. 5 to v. 6

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Update json files

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Update Accordion element

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Update Grid element to Grid v.2

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Update ListItem element to ListItemButton where deprecated button prop was used

veronikaslc committed:
[dependency] Update @mui/styles from v. 5 to v. 6

veronikaslc committed:
[dependency] Update @uiw/react-md-editor from v. 3 to v. 4

veronikaslc committed:
[dependency] Update css-loader from v. 6 to v. 7

veronikaslc committed:
[dependency] Update eslint from v. 8 to v. 9

veronikaslc committed:
[dependency] Update eslint-plugin-react to the latest release

veronikaslc committed:
[dependency] Update history from v. 4 to v. 5

veronikaslc committed:
[dependency] Update react-to-print from v. 2 to v. 3

veronikaslc committed:
[dependency] Update style-loader from v. 3 to v. 4

veronikaslc committed:
[dependency] Update uuid from v. 9 to v. 11

veronikaslc committed:
[dependency] Update yup from v. 0 to v. 1

veronikaslc committed:
[dependency] Bump node pom file dependency from v. 16 to v. 22

veronikaslc committed:
[dependency] Bump yarn pom file version

veronikaslc committed:
[dependency] Bump webpack and babel dev dependencies

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7

veronikaslc committed:
[dependency] Update React from v. 18 to v. 19
Refactored default params

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7
Providing location prop

veronikaslc committed:
[dependency] Minor UI dependency pruning
- Bumped recharts, cornerstone-wado-image-loader, react-number-format, react-google-autocomplete, google-palette dependency versions to the latest
- Removed unused semantic-ui-css, semantic-ui-react, set-blocking, formik-material-fields

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Grouped deprecated input props to slotProps for TextFiels

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Replaced deprecated props with slotsProps in various components

veronikaslc committed:
[misc] Fixed updating unmounted component

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7
Fixed form/subject/questionnaire page routing: separated rendering item page and table to a different components

veronikaslc committed:
[dependency] Update React from v. 18 to v. 19
Fixed grid and header alignments for Forms and Questionnaires

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7
Fixed Form rendering on save & view: force form checkout and re-ferch as useEffect(() => {}, []) is not triggered anymore if navigate() routs to the same component render

sdumitriu commented:
Subject's Timeline view doesn't work, null exception.

sdumitriu commented:
Switching forms between view and edit (and the other way around) doesn't work well, the form itself keeps its previous state.

sdumitriu commented:
Administration - Questionnaires - switching a questionnaire between edit and preview puts an `undefined` in the URL instead of the questionnaire name.

sdumitriu commented:
Administration - Groups - Add user to group dialog doesn't list the users.

sdumitriu commented:
Administration - Google API key -> Trying to change the key displays a JS error: `Error creating GoogleApiKey node: TypeError: location.reload is not a function`

sdumitriu commented:
Administration - Groups - Remove user -> JS error: `Uncaught TypeError: can't convert undefined to object
    handleRemoveUsers webpack-internal:///./src/Userboard/Groups/groupsmanager.jsx:111`

sdumitriu commented:
Patient Portal is broken:
```
Uncaught TypeError: /Forms\/([^.\/]+)/.exec(...) is null
    Form webpack-internal:///./src/questionnaire/Form.jsx:263
```

veronikaslc committed:
[dependency] UI dependency update
Fixed subject timeline navigation
Fixed switching a questionnaire between edit and preview puts an undefined in the URL instead of the questionnaire name
Fixed Add user to group dialog doesn't list the users
Fixed location reload for google API page

veronikaslc committed:
[dependency] UI dependency update
Fixed patient portal form navigation

veronikaslc committed:
[misc] Fixed JS error on removing user from group

veronikaslc commented:
Unsolved Routing problem: Switching forms between view and edit, visible on questionnaires like CPES in prems.

sdumitriu commented:
Clicking on links in the Quick Search results reloads the whole page instead of quick navigation.

sdumitriu commented:
This button needs to be centered:
![image](https://github.com/user-attachments/assets/b13ba074-032e-4cd1-8001-0d36883090ba)

sdumitriu commented:
This one too:
![image](https://github.com/user-attachments/assets/fb1f7a00-d400-49af-b799-94c360a0f829)

sdumitriu committed:
[misc] Various bugfixes related to switching between form's view and edit mode

veronikaslc committed:
[misc] Better routing for the Quick Search results links

veronikaslc committed:
[misc] Fix DOM errors of nested p components

veronikaslc committed:
[misc] Fixed Grid styling for patient survey

veronikaslc committed:
[misc] Fixed missing unique key warning

marta- commented:
Bug: broken layout for compact sections, e.g. sparc -> Clinician Assessment -> Page 4

![image](https://github.com/user-attachments/assets/616ff9f9-c339-4c54-86fe-e050764967c4)

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
Proper Grid v.2 props passing via size object

veronikaslc committed:
[misc] Better class name

sdumitriu commented on modules/patient-portal/src/main/frontend/src/patient-portal/SurveyInstructionsConfiguration.jsx:
```diff
@@ -78,7 +78,7 @@ function SurveyInstructionsConfiguration() {
         >
           <List className={classes.formEntries}>
-            { Object.keys(labels).map(category => { return (<>
+            { Object.keys(labels).map(category => { return (<span key={category + category}>
```
Why `category+category` and not `category+"Wrapper"`?

sdumitriu commented:
There are more things broken with forms not updating their content when the location changes. For example, in prems mode, create a visit information form with some value for the Location question, which will be copied over a few other forms. The quick search for that location, clicking on any of the forms will change the URL in the address bar, but the form doesn't change.

I think a quicker fix would be to separate the view from the actual form rendering component. Add a `FormView` component that is just responsible for extracting the ID and mode from the URL, then passes them as props to the child `Form` component. Since the props will change whenever the URL changes, the component will fully re-render.

Or, find more places that need the `id` to be taken into account.

sdumitriu committed:
[bugfix] QuestionMatrix answers don't work

veronikaslc committed:
[misc] Rendering Form via new shell FormView component to solve form routing issue

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -107,9 +107,10 @@ function Form (props) {
-  let [ isEdit, setIsEdit ] = useState(window.location.pathname.endsWith(".edit") || mode == "edit");
-  let [ isSummary, setIsSummary ] = useState(window.location.pathname.endsWith(".summary") || mode == "summary");
+
+  let id = useState(props.id || /Forms\/([^.\/]+)/.exec(location.pathname)[1]);
```
You don't need the `useState` call, just:
```
let id = props.id || /Forms\/([^.\/]+)/.exec(location.pathname)[1];
```
Same for the other two.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormView.jsx:
```diff
@@ -0,0 +1,40 @@
+  return (
+      <Form id={id} mode={mode}/>
+    );
+}
```
You need the `export default` call to make the component visible.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FormView.jsx:
```diff
@@ -0,0 +1,40 @@
+
+import React, { useEffect, useState } from "react";
+
+import { Form } from "./Form";
```
```suggestion
import Form from "./Form";
```

sdumitriu committed:
[misc] Fix react error when displaying a form

sdumitriu committed:
[misc] Partially fix forms not detecting the location change

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7
- Used unstable_HistoryRouter with history package for beforeunload event handling;
- Refactored Main component to be functional;

veronikaslc committed:
[dependency] Update React from v. 18 to v. 19
Proper way to set the defaults for props

veronikaslc committed:
[misc] Better class name

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/EditButton.jsx:
```diff
@@ -30,7 +30,7 @@ import QuestionnaireStyle from "../questionnaire/QuestionnaireStyle.jsx";
  */
 function EditButton(props) {
-  const { entryPath, entryType, size, className, admin, onClick } = { entryType: "", size: "large", ...props };
+  const { entryPath, entryType = "", size = "large", className, admin, onClick } = props;
```
why not on separate lines?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormPageNavigation.jsx:
```diff
@@ -42,7 +42,7 @@ import FormattedText from "../components/FormattedText";
  */
 function FormPageNavigation (props) {
-  const { pages, activePage, saveButton, backButton, isPageCompleted, navigateTo } = { activePage: 0, ...props };
+  const { pages, activePage = 0, saveButton, backButton, isPageCompleted, navigateTo } = props;
```
One per line

veronikaslc committed:
[dependency] Update React from v. 18 to v. 19
Proper way to set the defaults for props

veronikaslc committed:
[dependency] Update @mui/material from v. 5 to v. 6
- Proper Grid v.2 props passing via size object;
- Better alignments for Subject page elements;
- Better alignment for entities table action column to the right;
- Better center alignment for items in AdminDashboard;

veronikaslc committed:
[misc] Consistent styling of dashboard views loading state

veronikaslc committed:
[misc] Bugfix related to switching between form's view and mode

veronikaslc committed:
[dependency] Update React from v. 18 to v. 19
Proper way to set the defaults for props

veronikaslc committed:
[misc] Fix error of unknown onAddSuggestion event handler for TextField component

veronikaslc committed:
[misc] Fixed section header positioning

veronikaslc committed:
[misc] Closer spacing between action buttons in Livetable

sdumitriu committed:
[bugfix] 409 error when saving the same form twice

sdumitriu committed:
[misc] More generic detection of the form display mode

veronikaslc committed:
[misc] Added circular progress during form saving process

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -649,6 +650,15 @@ function Form (props) {
               />
             }
+            {saveInProgress &&
+              <span>
```
The span container is visibly pushing the content down while the Backdrop is displayed.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -268,6 +268,24 @@ const questionnaireStyle = theme => ({
     circularProgressContainer: {
         marginTop: theme.spacing(5),
     },
+    formCircularProgressContainer: {
+        position: "fixed",
+        backgroundColor: theme.palette.background.paper,
+        zIndex: theme.zIndex.drawer + 10,
+        top: "50%",
+        left: "50%",
+        [theme.breakpoints.up('md')]: {
+          marginLeft: "130px",
+        },
```
All this styling would not be needed if the CircularProgress was a child of the Backdrop.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -649,6 +650,15 @@ function Form (props) {
               />
             }
+            {saveInProgress &&
+              <span>
+                <Backdrop
+                  open={saveInProgress}
+                  className={classes.circularProgressBackdrop}
+                />
+                <CircularProgress className={classes.formCircularProgressContainer}/>
+              </span>
```
I would replace with:
```
             <Backdrop
               open={saveInProgress}
               sx={(theme) => ({
                 backgroundColor: alpha(theme.palette.background.paper, .5),
                 marginLeft: {md : "260px"},
                 zIndex: theme.zIndex.drawer + 1
               })}
             >
                <CircularProgress />
             </Backdrop>
```
Also remove the custom styling from QuestionnaireStyle.jsx and `import { alpha } from '@mui/material/styles';` at the top.

veronikaslc replied:
Much better! works perfectly

veronikaslc replied:
Fixed [here](37a620d540299980afe26c620d363de1db64bfd2)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
Shouldn't there have been a change that invalidates the data when saving is initiated?

veronikaslc replied:
Do you mean the one removed [here](https://github.com/data-team-uhn/cards/commit/2126b61d3c1c97afedc75d50c8a405207aecf85a)?

veronikaslc committed:
[misc] Added circular progress during form saving process
Better and more optimal styling

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -268,16 +268,6 @@ const questionnaireStyle = theme => ({
-          marginLeft: "130px",
-        },
-    },
     circularProgressBackdrop: {
```
Shouldn't this be removed as well?

veronikaslc replied:
Indeed, removed [2583682](https://github.com/data-team-uhn/cards/pull/1906/commits/258368237799f1fb936c24bc7e80bbcd7e0f94d8)

veronikaslc committed:
[misc] Added circular progress during form saving process
Removed leftover styling

marta- commented on modules/commons/src/main/frontend/src/components/ErrorDialog.jsx:
```diff
@@ -61,12 +61,19 @@ const useStyles = makeStyles(theme => ({
 
   return (
-    <Dialog onClose={onClose} {...rest}>
+    <Dialog onClose={onClose} title={title} maxWidth={maxWidth} fullWidth={fullWidth} {...rest}>
```
`Dialog` doesn't have a `title` prop
```suggestion
    <Dialog onClose={onClose} maxWidth={maxWidth} fullWidth={fullWidth} {...rest}>
```

veronikaslc committed:
[misc] Removed wrong title property from Dialog component

Co-authored-by: Marta Girdea <651980+marta-@users.noreply.github.com>

marta- commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -18,152 +18,137 @@
 import withStyles from '@mui/styles/withStyles';
 import GlobalStyles from '@mui/material/GlobalStyles';
-import { Redirect, Router, Route, Switch } from "react-router-dom";
+import { unstable_HistoryRouter as Router, Routes, Route, Navigate } from "react-router";
```
`unstable_HistoryRouter` - this is kind of unsettling...

veronikaslc replied:
The only known way we can use our own history to save form on page leave

marta- commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -18,152 +18,137 @@
-  switchRoutes = (routes) => {
-    return (<Switch color="secondary">
+  let switchRoutes = () => {
+    return (<Routes color="secondary">
```
This `color="secondary"` must be an artifact from setting the color to all our `MUI` switches a while back. Should be removed from all the `router` components.

marta- commented on modules/homepage/src/main/frontend/src/themePage/index.jsx:
```diff
@@ -187,11 +172,11 @@ root.render(
-          <Redirect from="/" to="/content.html/Questionnaires/User"/>
-          <Redirect from="/content" to="/content.html/Questionnaires/User" />
-        </Switch>
+        <Routes color="secondary">
```
Another `color="secondary"` to remove.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx:
```diff
@@ -708,30 +707,28 @@ function QuestionnaireSet(props) {
 
   let submitButton = (label) => (
-    <Fab variant="extended" disabled={submissionInProgress} color="primary" onClick={() => {onSubmit()}} key="review-submit">
+    <Fab variant="extended" disabled={submissionInProgress} color="primary" onClick={() => {onSubmit()}} key={"review-submit"+label}>
```
Why did we need to add `+label`?

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/QuestionnaireSet.jsx:
```diff
@@ -759,8 +756,8 @@ function QuestionnaireSet(props) {
   let finalInstructions = (
-      endingMessage ? <FormattedText paragraph key="summary-instructions">{endingMessage}</FormattedText> :
-      displayText("summaryInstructions", FormattedText, {color: "textSecondary", key: "summary-instructions", paragraph: true})
+      endingMessage ? <FormattedText key="summary-instructions">{endingMessage}</FormattedText> :
+      displayText("summaryInstructions", FormattedText, {color: "textSecondary", key: "summary-instructions"})
```
Shouldn't `paragraph` have been replaced by `component="p"` instead of removing it completely?

veronikaslc replied:
This is intentional for better positioning.

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/SurveyInstructionsConfiguration.jsx:
```diff
@@ -78,7 +78,7 @@ function SurveyInstructionsConfiguration() {
         >
           <List className={classes.formEntries}>
-            { Object.keys(labels).map(category => { return (<>
+            { Object.keys(labels).map(category => { return (<span key={category + "Wrapper"}>
```
By adding this span, we're wrapping block elements in an inline element, resulting in invalid html. Instead, put the key on the `Fragment`:
```suggestion
            { Object.keys(labels).map(category => { return (<React.Fragment key={category + "Wrapper"}>
```

marta- commented on modules/patient-portal/src/main/frontend/src/patient-portal/index.jsx:
```diff
@@ -97,11 +97,11 @@ root.render(
-          <Redirect from="/Survey" to="/Survey.html/"/>
-          <Redirect from="/" to="/Survey.html/"/>
-        </Switch>
+        <Routes color="secondary">
```
Remove `color="secondary"`

marta- commented:
Shouldn't all the `.propTypes` have been removed in addition to `.defaultProps`?

veronikaslc committed:
[misc] Removed artifact color property from Routes

veronikaslc committed:
[misc] Wrapping block in React Fragment element

veronikaslc commented:
Left them for our own reference.

> Shouldn't all the `.propTypes` have been removed in addition to `.defaultProps`?

Left intensionally for some discussion. 

The short answer is "yes", however there is something we can do to avoid the chaos when there's no more specification of what each prop type is expected to be. Here is what they propose in a PR in React repo where they [Remove JSX propTypes validation](https://github.com/facebook/react/pull/28328).

> In general, our recommendation is to use static type checking (e.g. TypeScript). If you'd like to still run propTypes checks, you can do so manually, same as you'd do outside React:
> 
> ```js
> import checkPropTypes from 'prop-types/checkPropTypes';
> 
> function Button(props) {
>   checkPropTypes(Button.propTypes, props, 'prop', Button.name)
>   // ...
> }
> ```
> This could be automated as a Babel plugin if you want to keep these checks implicit. (We will not be providing such a plugin, but someone in community might be interested in building or maintaining one.)

Wdyt?

veronikaslc commented:
- New [babel-plugin-check-prop-types](https://www.npmjs.com/package/babel-plugin-check-prop-types) is used for implicit propTypes checks for `development` mode.
- Old [babel-plugin-transform-react-remove-prop-types](https://www.npmjs.com/package/babel-plugin-transform-react-remove-prop-types) is used to remove `propTypes` and `prop-types` import statements from the `production` build.

veronikaslc committed:
[dependency] Update @emotion dependencies to the latest v. 11.14.0

veronikaslc committed:
[misc] Run explicit propTypes check

veronikaslc committed:
[dependency] Locking react-is for v.19 for prop-types proper check of PropTypes.Element

veronikaslc committed:
[misc] Fixed check prop types warnings

veronikaslc commented:
The problem left to fix:
- open the form (view mode) in two tabs
- edit, change and save in one tab
- edit in the second tab -> does it load the correct data?

veronikaslc committed:
[misc] Better loading/saving state visualisation via loading spinner

veronikaslc committed:
[misc] Fixed check prop types warnings
Remove .Required from props that have defaults and give warnings
Add missing formatted displayMode to TextQuestion

veronikaslc committed:
[dependency] Update react-router from v. 5 to v. 7
Fixed data component rendering on view/edit mode change via conditional rendering FormEntry on top level in Form

veronikaslc committed:
Merge branch 'dev' into ui-dependency-update

veronikaslc committed:
Merge branch 'dev' into ui-dependency-update

veronikaslc committed:
CARDS-2653: Flash content change when closing the Edit Subject Type dialog

sdumitriu committed:
Merge pull request #1934 from data-team-uhn/CARDS-2653

CARDS-2653: Flash content change when closing the Edit Subject Type dialog
